prql-elixir/native/prql/src/lib.rs

```diff
@@ -0,0 +1,161 @@
+            result: sql,
+        }),
+        Err(e) => Ok(Response {
+            status: atoms::error(),
+            result: e.to_json(),
+        }),
+    }
+}
+
+/// Get the dialect from an atom. By default `Generic` dialect will be used
+fn dialect_from_atom(a: Atom) -> prql_compiler::sql::Dialect {
+    use prql_compiler::sql::Dialect as D;
+
+    if a == atoms::ansi() {
+        D::Ansi
+    } else if a == atoms::big_query() {
+        D::BigQuery
+    } else if a == atoms::click_house() {
+        D::ClickHouse
+    } else if a == atoms::generic() {
+        D::Generic
+    } else if a == atoms::hive() {
+        D::Hive
+    } else if a == atoms::mssql() {
+        D::MsSql
+    } else if a == atoms::mysql() {
+        D::MySql
+    } else if a == atoms::postgres() {
+        D::PostgreSql
+    } else if a == atoms::sql_lite() {
+        D::SQLite
+    } else if a == atoms::snow_flake() {
+        D::Snowflake
+    } else {
+        D::Generic
+    }
+}
+
+impl From<CompileOptions> for prql_compiler::sql::Options {
+    /// Get `prql_compiler::sql::Options` options from `CompileOptions`
+    fn from(o: CompileOptions) -> Self {
+        prql_compiler::sql::Options {
+            format: o.format,
+            dialect: Some(dialect_from_atom(o.dialect)),
+            signature_comment: o.signature_comment,
+        }
+    }
+}
+
+#[derive(Clone, NifStruct, Debug)]
+#[module = "PRQL.Native.CompileOptions"]
+pub struct CompileOptions {
+    /// Pass generated SQL string trough a formatter that splits it
+    /// into multiple lines and prettifies indentation and spacing.
+    ///
+    /// Defaults to true.
+    pub format: bool,
+
+    /// Target dialect you want to compile for.
+    ///
+    /// Because PRQL compiles to a subset of SQL, not all SQL features are
+    /// required for PRQL. This means that generic dialect may work with most
+    /// databases.
+    ///
+    /// If something does not work in dialect you need, please report it at
+    /// GitHub issues.
+    ///
+    /// If None is used, `sql_dialect` flag from query definition is used.
+    /// If it does not exist, [Dialect::Generic] is used.
+    pub dialect: Atom,
+
+    /// Emits the compiler signature as a comment after generated SQL
+    ///
+    /// Defaults to true.
+    pub signature_comment: bool,
+}
+
+#[derive(NifTuple)]
+pub struct Response {
+    /// status atom `:ok` or `:error`
+    status: Atom,
+
+    /// result string
+    result: String,
+}
+
+#[rustler::nif]
+/// compile a prql query into sql
+pub fn compile(prql_query: &str, options: CompileOptions) -> NifResult<Response> {
+    to_result_tuple(
+        Ok(prql_query)
+            .and_then(prql_compiler::prql_to_pl)
+            .and_then(prql_compiler::pl_to_rq)
+            .and_then(|rq| {
+                prql_compiler::rq_to_sql(rq, Some(options).map(prql_compiler::sql::Options::from))
+            })
+            .map_err(|e| e.composed("", prql_query, false)),
+    )
+}
```

Comment:
For these functions, would it be possible to instead just defer to the equivalent function upstream?

We'd need to convert the `CompileOptions` struct at the start, and the output at the end, but that would reduce the duplication for most of this, and mean that upstream changes were reflected here immediately.

Reply:
@max-sixty i didn't get what you meant by `upstream` changes

Reply:
Sorry, by `upstream`, I mean the code in `prql-compiler`.

So rather than:

```rust
pub fn compile(prql_query: &str, options: CompileOptions) -> NifResult<Response> {
    to_result_tuple(
        Ok(prql_query)
            .and_then(prql_compiler::prql_to_pl)
            .and_then(prql_compiler::pl_to_rq)
            .and_then(|rq| {
                prql_compiler::rq_to_sql(rq, Some(options).map(prql_compiler::sql::Options::from))
            })
            .map_err(|e| e.composed("", prql_query, false)),
    )
}
```

could we do something like

```rust
pub fn compile(prql_query: &str, options: CompileOptions) -> NifResult<Response> {
    let rust_options = rust_options_of_ex_options(options);
    let result = prql_compiler::compile(prql_query, rust_options);
    to_result_tuple(result)
}
```

(this is approximate, but I hope it communicates the point; please let me know if it's unclear!)

That way, if the code like `.and_then(prql_compiler::prql_to_pl)` changes upstream / in `prql-compiler`, this library takes the changes without having to change its code.

Reply:
Got it, i was following the JS example but this makes sense

Reply:
Ah you're right, that's not great. 

Let me add a TODO to that code to change it. Sorry for the misdirection.

Reply:
Thanks @kasvith https://github.com/PRQL/prql/pull/1504

Reply:
@max-sixty shall we continue in discord? i ve some doubts to be cleared out

Reply:
Updated